VultureProject/darwin
# Version 1.2

# Changelog 1.2

## BREAKING

- New python dependency **jsonschema 3.2.0**
- New python dependency **psutil 5.6.6**
- The filters no longer launch alert on 101 code (error code).

This version enables each filter to manage its alerts independently. This make flogs obsolete.
However, to enable smoother transition, the old configuration and flogs are still perfectly working.
It is important to note that **with new alerting configuration an alert will be raised even if the client sets `response_type` to `back` or `no`**.

**NOTE: flogs and old filter configuration will become obsolete and will no longer be supported in Version 1.3.**

## CMake
The CMakeList is now only in charge of calling .cmake files for each filter it needs to compile.
**The behavior stays the same.**

## Manager
- New configuration format created for the darwin manager (darwin.conf). Legacy configuration is still working.
  - possibility to add objects in json for general manager/Darwin mechanics
  - no more variable object names in conf -> filters are self-contained in a list, with the name as a regular field
- Fix monitor crashing when request misses a field
- Implement reporter thread, to report stats at regular intervals on file/redis
- Improve management on threads on stop with a condition variable
- Implement a function to get informations on a running process

## Core
- Implemented Statistics
  - status moved to Stats class
  - added 'stopping' to statuses
  - number of connections
  - number of entries (lines) received
  - number or incorrect entries
  - number of matches (certitude >= threshold)
- Rework of FileManager to ensure better error handling.
- Adding an AlertManager to the Core to enable every filter to raise an alert. The alerts can be raised in : File, Redis list, Redis channel.
- Alert file rotation is handled catching the SIGHUP or SIGUSR1 signal.
- The new alerting configuration fields are:
  - 'log_file_path' _(optional)_: path to the log file dedicated to alerts
  - 'redis_socket_path' _(optional)_: path to the redis UNIX socket to send alert to
  - 'alert_redis_list_name' _(optional)_: name of the alert list in redis
  - 'alert_redis_channel_name' _(optional)_: name of redis channel to publish alert to

_NOTE: If the alert configuration is missing or wrong, it is not applied. If the alert configuration is partialy correct, apply as much as possible. In both cases a warning is logged._

## Filters
- Filters are now using the AlertManager to raise alert.
- Added ftest to enable easier testing of the core functions. This filter is not compiled by default but is mandatory for tests.

## Tests
### Test Tools
- RedisServer able to start a Redis server with a listening IP AND a UNIX socket

### Core
- Added the alert manager tests (around 75 tests so no listing here)

### FTAnomaly
- well_formatted_data_test
- data_too_short_ignored_test
- not_data_string_ignored_test
- not_data_list_ignored_test
- invalid_ip_ignored_test
- invalid_protocol_ignored_test
- thread_working_test (**CAUTION** this test wait **5 minutes** to see if the thread really take redis data)
- verify if received alert in channel
- verify if received alert in list
- verify if received alert in file
- verify data stored in list (can have same name that the channel)
- verify if no alert in list with same name
- alert in redis test
- alert published test
- alert in file test
- check for proper handling of both legacy and new configuration format


### Manager
- Configuration v2 tests added
- Retro-compatibility tests added
- Stats reporting tests added
- Test for filters update with no filter added
